xdp/src/route_monitor.rs

```diff
@@ -0,0 +1,182 @@
+use {
+    crate::{
+        netlink::{
+            is_supported_ipv4_neigh_header, is_supported_ipv4_route_header, parse_rtm_newneigh,
+            parse_rtm_newroute, NetlinkMessage, NetlinkSocket,
+        },
+        route::Router,
+    },
+    arc_swap::ArcSwap,
+    libc::{
+        self, pollfd, POLLERR, POLLHUP, POLLIN, POLLNVAL, RTMGRP_IPV4_ROUTE, RTMGRP_NEIGH,
+        RTM_DELNEIGH, RTM_DELROUTE, RTM_NEWNEIGH, RTM_NEWROUTE,
+    },
+    log::*,
+    std::{
+        io::{Error, ErrorKind},
+        net::IpAddr,
+        sync::{
+            atomic::{AtomicBool, Ordering},
+            Arc,
+        },
+        thread,
+        time::{Duration, Instant},
+    },
+};
+pub struct RouteMonitor;
+
+impl RouteMonitor {
+    /// Subscribes to RTMGRP_IPV4_ROUTE | RTMGRP_NEIGH multicast groups
+    /// Waits for updates to arrive on the netlink socket
+    /// Batches updates for `drain_window`
+    /// Once drain_window expires, publishes the working routing table to the atomic router
+    /// If the netlink socket is invalid, recreates the socket and resyncs the atomic router
+    pub fn start(
+        atomic_router: Arc<ArcSwap<Router>>,
+        exit: Arc<AtomicBool>,
+        update_interval: Duration,
+    ) -> thread::JoinHandle<()> {
+        thread::Builder::new()
+            .name("solRouteMon".to_string())
+            .spawn(move || {
+                let mut state =
+                    RouteMonitorState::new(Router::new().expect("error creating Router"));
+
+                let timeout = Duration::from_millis(10);
+                while !exit.load(Ordering::Relaxed) {
+                    state.publish_if_needed(&atomic_router, update_interval);
+
+                    let mut pfd = pollfd {
+                        fd: state.sock.as_raw_fd(),
+                        events: POLLIN,
+                        revents: 0,
+                    };
+
+                    let ev = match poll(&mut pfd, timeout) {
+                        // timeout
+                        Ok(0) => continue,
+                        Ok(_) => pfd.revents,
+                        Err(e) => {
+                            error!("netlink poll error: {e}");
+                            state.reset(&atomic_router);
+                            continue;
+                        }
+                    };
+
+                    assert!(ev & POLLNVAL == 0);
```

Comment:
in light of the current cloduflare fiasco, please put this back as a
debug_assert 😂